[Draft] Slash malevolent validators

akildemir opened the pull request:
This is a **draft** pr to slash validators with bad behavior. Comments and/or reviews are appreciated.

in tributary, tendermint is used for the consensus. And there are bunch of defined invalid or bad behaviors for validators that participate in the tendermint consensus algorithm. We slash those nodes to disincentive those behaviors.

In this case, once a node catches an invalid/bad behavior from another validator, it either sends an evidence for this bad  behavior or a vote to slash the bad node. If an evidence provided and valid, it is solely enough to slash the node by the network participants even if other nodes didn't see this behavior. Once the evidence is on the chain, the bad node is slashed. If a vote is sent, then we wait vote from 2/3 of the nodes to slash the node.

For now following points are completed;

- create a special tx type to publish bad validator behavior data and/or votes for slashing
- modify the tributary to hold the both app txs and this integral tx
- ignore unsigned tendermint tx replays
- verify that provided evidence was actually valid(sender isn't falsely try to claim someone else with invalid evidence).
- scan the slash evidence/votes in the chain
- support unsigned app txs while at it(not really related to this pr but hey! why not)
- enforce provided, unsigned, signed tx ordering within a block
- add unit tests for tendermint tx validation
- update existing mempool/blockchain tests for new tx types

Commit:
add slash tx

Commit:
ignore unsigned tx replays

Commit:
verify that provided evidence is valid

Commit:
fix clippy + fmt

Commit:
move application tx handling to another module

Commit:
partially handle the tendermint txs

kayabaNerve commented on coordinator/tributary/src/blockchain.rs:
```diff
@@ -65,17 +69,33 @@ impl<D: Db, T: TransactionTrait> Blockchain<D, T> {
       block_number: 0,
       tip: genesis,
       next_nonces,
+      unsigned_included: vec![],
```
Why is this here?

1) The mempool checks it for a TX's existence. That means it should be a HashSet, not a vec, if kept.
2) We shouldn't store in RAM a full list of every unsigned TX of all time. This should remain at the DB level. Pass the Mempool a lambda, or have it wrap its DB handle with BlockchainDb to get access to the function in BcDb.

kayabaNerve commented on coordinator/tributary/src/blockchain.rs:
```diff
@@ -173,7 +193,21 @@ impl<D: Db, T: TransactionTrait> Blockchain<D, T> {
+          existing.extend(hash);
+          txn.put(key, existing);
+
+          // add to memory
```
If it's in the DB, it shouldn't also be in memory.

kayabaNerve commented on coordinator/tributary/src/mempool.rs:
```diff
@@ -45,6 +46,18 @@ impl<D: Db, T: TransactionTrait> Mempool<D, T> {
 
+  fn unsigned_already_exist(&self, unsigned_included: &[[u8; 32]], hash: [u8; 32]) -> bool {
+    // check we have the tx in the pool/chain
+    let res = unsigned_included.iter().find(|&&tx_hash| {
```
Should check the DB, not memory.

kayabaNerve reviewed (commented):
Just a few comments while I had a second. I don't think GH is showing me the full edits, so I'll have to re-open this to make the rest of my comments...

kayabaNerve commented on coordinator/src/tributary/db.rs:
```diff
@@ -27,6 +27,25 @@ impl<D: Db> TributaryDb<D> {
+
+  pub fn set_fatally_slashed(txn: &mut D::Transaction<'_>, genesis: [u8; 32], id: [u8; 32]) {
+    let key = Self::fatal_slash_key(genesis);
+    let mut existing = txn.get(&key).unwrap_or(vec![]);
```
This should check if the entity is already present.

Also, what's ID? Don't we want to slash keys, represented by Ristretto::G?

akildemir replied:
ID is the ValidatorId from `msg.sender` which I believe is already Ristretto::G.

kayabaNerve commented on coordinator/src/tributary/db.rs:
```diff
@@ -27,6 +27,25 @@ impl<D: Db> TributaryDb<D> {
+    Self::tributary_key(b"slash_point", [genesis, id].concat())
+  }
+
+  pub fn slash_vote_key(genesis: [u8; 32], id: [u8; 32]) -> Vec<u8> {
```
We do vote on slashes by ID, yet we also need to track voted on slash ID -> underlying key.

akildemir replied:
you mean who sent this vote? IF so, that is already a TODO atm.

akildemir replied:
you mean who sent this vote? If so, that is a TODO atm.

kayabaNerve commented on coordinator/src/tributary/db.rs:
```diff
@@ -27,6 +27,25 @@ impl<D: Db> TributaryDb<D> {
+    Self::tributary_key(b"fatal_slash", genesis)
+  }
+
+  pub fn slash_point_key(genesis: [u8; 32], id: [u8; 32]) -> Vec<u8> {
```
Shouldn't this be Ristretto::G?

kayabaNerve commented on coordinator/tributary/src/block.rs:
```diff
@@ -146,12 +152,14 @@ impl<T: Transaction> Block<T> {
     &self,
     genesis: [u8; 32],
     last_block: [u8; 32],
     mut locally_provided: HashMap<&'static str, VecDeque<T>>,
```
We also need to verify unsigned TXs aren't being included multiple times.

Preferably, we enforce an order of Provided, Unsigned, Signed as well.

kayabaNerve commented on coordinator/tributary/src/blockchain.rs:
```diff
@@ -102,16 +129,42 @@ impl<D: Db, T: Transaction> Blockchain<D, T> {
+  pub(crate) fn commit_by_block_number(&self, block: u32) -> Option<Vec<u8>> {
+    let hash = self.block_hash(block);
+    hash.as_ref()?;
+    self.commit(hash.unwrap().as_slice().try_into().unwrap())
```
This unwrap can be replaced with `?`, removing the prior line as well.

kayabaNerve replied:
Also, why not have block_hash return [u8; 32]?

kayabaNerve commented on coordinator/tributary/src/blockchain.rs:
```diff
@@ -102,16 +129,42 @@ impl<D: Db, T: Transaction> Blockchain<D, T> {
-  pub(crate) fn add_transaction(&mut self, internal: bool, tx: T) -> bool {
-    self.mempool.add(&self.next_nonces, internal, tx)
+  pub(crate) fn add_transaction<N: Network>(&mut self, internal: bool, tx: Transaction<T>, schema: N::SignatureScheme) -> bool {
+
```
Random new line.

kayabaNerve commented on coordinator/tributary/src/blockchain.rs:
```diff
@@ -102,16 +129,42 @@ impl<D: Db, T: Transaction> Blockchain<D, T> {
+      let hash = Self::block_hash_from_db(db, genesis, block);
+      hash.as_ref()?;
+      // we must have a commit per valid hash
+      let commit = Self::commit_from_db(db, genesis, hash.unwrap().as_slice().try_into().unwrap()).unwrap();
```
Same comment on block_hash handling as prior comment discussing it.

kayabaNerve commented on coordinator/tributary/src/blockchain.rs:
```diff
@@ -156,10 +217,12 @@ impl<D: Db, T: Transaction> Blockchain<D, T> {
     self.tip = block.hash();
     txn.put(self.tip_key(), self.tip);
-
+    
```
Random tab.

kayabaNerve commented on coordinator/tributary/src/mempool.rs:
```diff
@@ -4,108 +4,165 @@ use ciphersuite::{Ciphersuite, Ristretto};
+              }
+              res.txs.insert(hash, Transaction::Application(tx));
+            }
+            _ => panic!("mempool database had a non-signed application transaction"),
```
Unsigned transactions are now mempool routed, so we need to handle Transaction::Unsigned and update this error to "non-provided".

kayabaNerve replied:
@akildemir Application transactions can be unsigned and are allowed in the mempool so long as they're valid.

kayabaNerve commented on coordinator/tributary/src/mempool.rs:
```diff
@@ -4,108 +4,165 @@ use ciphersuite::{Ciphersuite, Ristretto};
+
+        // check we have the tx in the pool/chain
+        if self.unsigned_already_exist(unsigned_included, tx.hash()) {
+          // TODO: should this return true? function comment says return false
```
Returning false is correct.

kayabaNerve commented on coordinator/tributary/src/mempool.rs:
```diff
@@ -118,27 +175,31 @@ impl<D: Db, T: Transaction> Mempool<D, T> {
-          if blockchain_next_nonces[signer] > *nonce {
-            self.remove(&hash);
-            continue;
+      if let Transaction::Application(tx) =  tx {
```
`=  tx`? There's two spaces there.

kayabaNerve replied:
This still has an extra space.

kayabaNerve commented on coordinator/tributary/src/mempool.rs:
```diff
@@ -118,27 +175,31 @@ impl<D: Db, T: Transaction> Mempool<D, T> {
+              continue;
+            }
           }
+          _ => panic!("non-signed transaction entered mempool"),
```
Same commentary on above about handling unsigned TXs here. We just need to check they weren't already included.

akildemir replied:
Don't we already check whether a given unsigned is already in chain when adding to the pool? why do we need to re check here?

kayabaNerve replied:
This block is checking if the transaction didn't go stale (included on chain). It has to be run on all TXs, not just application TXs. Unsigned TXs are stale if they were already included on chain.

Technically, if they're included on chain, they should be removed from the mempool and accordingly should never be present in the mempool if they're stale. This is distinct from Signed where another TX with the same nonce may enter, making the TX in our mempool stale, despite not being on-chain.

The reason I still want you to handle unsigned TXs here is:
1) It's assuming a lack of a race condition which is exceptionally easy to add here. I'd rather be robust.
2) You iterate over all TXs. If any unsigned TX is in the pool, this will panic. While you limited it to Application TXs, so it wouldn't panic on unsigned Tendermint TXs, application TXs are allowed to be unsigned. That means an unsigned application TX will hit this panic.

kayabaNerve commented on coordinator/tributary/src/mempool.rs:
```diff
@@ -118,27 +175,31 @@ impl<D: Db, T: Transaction> Mempool<D, T> {
       }
 
       // Since this TX isn't stale, include it
+      // TODO: This also includes non-signed Tendermint Evidence txs. should it?
```
Yes, yet preferably in a seperate vec so we can return `unsigned || signed`, and enforce that ordering (unsigned before signed, just as we currently have provided before signed).

akildemir replied:
why do we have that ordering? can't a block contain all 3 types at the same time? or you mean the tx ordering within the same block?

kayabaNerve replied:
Within a block, ideally the transactions are ordered

Provided
Unsigned
Signed

kayabaNerve reviewed (commented):
I reviewed about half of this. Looks solid so far.

Commit:
fix pr comments

Commit:
support unsigned app txs

Commit:
add slash target to the votes

Commit:
enforce provided, unsigned, signed tx ordering within a block

Commit:
bug fixes

Commit:
add unit test for tendermint txs

Commit:
bug fixes

Commit:
update tests for tendermint txs

Commit:
add tx ordering test

Commit:
tidy up tx ordering test

kayabaNerve committed:
cargo +nightly fmt

kayabaNerve committed:
Misc fixes from rebasing

kayabaNerve committed:
Finish resolving clippy

kayabaNerve committed:
Remove sha3 from tendermint-machine

kayabaNerve committed:
Resolve a DoS in SlashEvidence's read

Also moves Evidence from Vec<Message> to (Message, Option<Message>). That
should meet all requirements while being a bit safer.

kayabaNerve committed:
Make lazy_static a dev-depend for tributary

kayabaNerve committed:
Various small tweaks

One use of sort was inefficient, sorting unsigned || signed when unsigned was
already properly sorted. Given how the unsigned TXs were given a nonce of 0, an
unstable sort may swap places with an unsigned TX and a signed TX with a nonce
of 0 (leading to a faulty block).

The extra protection added here sorts signed, then concats.

kayabaNerve committed:
Fix Tributary tests I broke, start review on tendermint/tx.rs

kayabaNerve committed:
Finish reviewing everything outside tests and empty_signature

kayabaNerve committed:
Remove empty_signature

empty_signature led to corrupted local state histories. Unfortunately, the API
is only sane with a signature.

We now use the actual signature, which risks creating a signature over a
malicious message if we have ever have an invariant producing malicious
messages. Prior, we only signed the message after the local machine confirmed
it was okay per the local view of consensus.

This is tolerated/preferred over a corrupt state history since production of
such messages is already an invariant. TODOs are added to make handling of this
theoretical invariant further robust.

kayabaNerve committed:
Remove async_sequential for tokio::test

There was no competition for resources forcing them to be run sequentially.

kayabaNerve committed:
Modify block order test to be statistically significant without multiple runs

kayabaNerve committed:
Clean tests
